File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode3000MaximumAreaOfLongestDiagonalRectangle.cpp
#include "LeetCode3000MaximumAreaOfLongestDiagonalRectangle.h"

#include <algorithm>
using namespace std;

int LeetCode3000MaximumAreaOfLongestDiagonalRectangle::areaOfMaxDiagonal(vector<vector<int>>& dimensions)
{
    int maxArea{};
    int maxDiagonal{};
    int diagonal{};
    int area{};
    for (const auto& dimension : dimensions)
    {
        diagonal = dimension[0] * dimension[0] + dimension[1] * dimension[1];
        if (diagonal == maxDiagonal)
        {
            area = dimension[0] * dimension[1];
            maxArea = max(maxArea, area);
            continue;
        }

        if (diagonal > maxDiagonal)
        {
            maxDiagonal = diagonal;
            maxArea = dimension[0] * dimension[1];;
        }
    }
    return maxArea;
}

File: LeetCodeProblemsGo/LeetCode3000MaximumAreaOfLongestDiagonalRectangle/LeetCode3000MaximumAreaOfLongestDiagonalRectangle.go
package LeetCode3000MaximumAreaOfLongestDiagonalRectangle

func areaOfMaxDiagonal(dimensions [][]int) int {
	maxDiagonal := 0
	diagonal := 0
	area := 0
	maxArea := 0
	for _, dimension := range dimensions {
		diagonal = dimension[0]*dimension[0] + dimension[1]*dimension[1]
		if diagonal == maxDiagonal {
			area = dimension[0] * dimension[1]
			maxArea = max(area, maxArea)
			continue
		}
		if diagonal > maxDiagonal {
			maxDiagonal = diagonal
			maxArea = dimension[0] * dimension[1]
		}
	}
	return maxArea
}
